Asana: Added start date functionality for Task creation

stanty522 opened the pull request:
## Description

I added the functionality to add an optional start date for tasks created for Asana. This allows for tasks to show up in Asana with a date range for completion. 

An example of this is if a task is created with a start date of Jan 1, 2023 and a due date of Jan 15, 2023 then on Asana it will show that this task has a range of dates where it can be completed. Whereas if only a due date was inputted then the task will only show up as a task to be completed by that date.

## Screencast

<!-- If you add a new extension or command, include a screencast (or screenshot for straightforward changes). A good screencast will make the review much faster - especially if your extension requires registration in other services.  -->

## Checklist

- [ ✅] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [ ✅] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [ ✅] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [ ✅] I checked that files in the `assets` folder are used by the extension itself
- [✅ ] I checked that assets used by the `README` are placed outside of the `metadata` folder

thomaslombart reviewed (changes requested):
Hey there 👋 
Thank you for your contribution!

I'm getting this error while trying the start date feature:

<img width="1000" alt="asana 2023-01-16 at 08 16 27" src="https://user-images.githubusercontent.com/16003285/212620014-e5732457-f68c-4647-9d93-64f544464017.png">

Note that I don't have any "start date" field on my workspace's task details, so maybe the PR is referring to a custom field of yours?

<img width="682" alt="CleanShot 2023-01-16 at 08 22 29@2x" src="https://user-images.githubusercontent.com/16003285/212620264-9c88347f-b6c7-44ae-ad68-b4e66b1a984d.png">

stanty522 commented:
I just did some googling and the start date feature is only available for Premium users. I confirmed this by testing the extension on my work Asana account which is on the Premium version and the personal Asana account. 

I think a basic if statement check here would work to just only show this to Premium asana account users.

thomaslombart commented:
That would work indeed 😊

stanty522 commented:
Hey I'm kind of stuck here. Maybe you can help. 

Here's the code I've written based on the above. And as far as I can tell, it should work. 

I've added this to CreateTaskForm.tsx

  ```
    {workspaces?.map((workspace) => {
      if (workspace.is_organization == true) {
              return (
                <Form.DatePicker title="Start Date" type={Form.DatePicker.Type.Date} {...itemProps.start_date} />
              );
            }})}
```

I've also added the "is_organization: boolean;" in workspaces.ts

Given the above, I expect that the if statement will return true for the Premium account and will return false for those that are not Premium.

I even double checked the api endpoint and it does return "true" for my Premium account and "false" for my free account. However when I run this code, its not showing the form element regardless if I'm signed on to my premium or free account.

thomaslombart commented:
Where did you add this code? It's hard to tell what's wrong without having the full context.

On the global approach, I would recommend not map on all of the workspaces, but rather retrieving the selected workspace (using `values.workspace` and the list of workspaces) to then conditionally show the field based on this workspace:

```tsx
const selectedWorkspace = workspaces.find(/* ... */)

// ...

{selectedWorkspace.is_organization ? /* Start date field */ : null}
```

stanty522 commented:
I added the selectedWorkspace variable on line 145, right after the hasCustomFields variable

 ```
 const selectedWorkspace = workspaces?.filter((workspace) => {
    return values.workspace?.includes(workspace.gid);
  });
```
Using selectedWorkspace I have the following code on line 198 which is just after the code for user assignee and just before the due date form.

```
      {selectedWorkspace?.map((workspace) => {
      if (workspace.is_organization === true) {
              return (
                <Form.DatePicker title="Start Date" type={Form.DatePicker.Type.Date} {...itemProps.start_date} />
              );
            }})}
```

This is all in the CreateTaskForm.tsx file.

For the workspaces.ts file I added the is_organization as part of the data payload

```
import { request } from "./request";

export type Workspace = {
  gid: string;
  name: string;
  is_organization: boolean;
};

export async function getWorkspaces() {
  const { data } = await request<{ data: Workspace[] }>("/workspaces");

  return data.data;
}

```

thomaslombart commented:
Hey @stanty522, could you push the code to the PR? It'll be easier for me to debug this.

stanty522 commented:
I added 2 commits - have a look!

thomaslombart commented:
Thanks, I'll have a look at that tomorrow 😊

thomaslombart commented:
Hey, so there were some problems in the code, indeed.

1st problem: there's only one workspace that can be selected at a time, so `filter` should be replaced by `find`, and there shouldn't be any `map` in the JSX code:

```ts
const selectedWorkspace = workspaces?.find((workspace) => values.workspace === workspace.gid);
```

And later in the TSX code:

```tsx
      {selectedWorkspace?.is_organization ? (
        <Form.DatePicker title="Start Date" type={Form.DatePicker.Type.Date} {...itemProps.start_date} />
      ) : null}
```

2nd problem: you need to specify which fields you want to get in Asana's APIs. While the API returns a default set of fields, it's common to ask Asana for extra fields:

```ts
export async function getWorkspaces() {
  const { data } = await request<{ data: Workspace[] }>("/workspaces", {
    params: { opt_fields: "gid,name,is_organization" },
  });

  return data.data;
}
```

stanty522 commented:
Thanks for the above! Have made the changes now. Sorry took a while as it was a busy last couple of weeks for me. Works as intended now.

thomaslombart commented:
No worries. I'll take a look at it during the day and hopefully merge the PR 😊

thomaslombart commented on extensions/asana/src/api/workspaces.ts:
```diff
@@ -3,10 +3,14 @@ import { request } from "./request";
 export type Workspace = {
   gid: string;
   name: string;
+  email_domains: string[];
```
This is probably not wanted.

thomaslombart reviewed (approved):
Looks good to me! There's only one small line to remove 😊

stanty522 commented:
Done! Made the change!
